crankshaft-docker/src/service.rs

```diff
@@ -0,0 +1,282 @@
+use futures::StreamExt;
+use futures::TryStreamExt;
+use tokio::time::sleep;
+use tracing::debug;
+
+use crate::Error;
+use crate::Result;
+
+/// A docker service.
+///
+/// Docker services are used to run tasks when Docker is configured to use a
+/// swarm.
+///
+/// This allows the swarm manager to schedule the task on available resources.
+///
+/// The service will always have a single replica of the task and the task will
+/// not restart.
+pub struct Service {
+    /// A reference to the [`Docker`] client that will be used to create this
+    /// service.
+    client: Docker,
+
+    /// The name of the service.
+    id: String,
+
+    /// Whether or not standard output is attached.
+    attach_stdout: bool,
+
+    /// Whether or not standard output is attached.
+    attach_stderr: bool,
+}
+
+impl Service {
+    /// Creates a new [`Service`] if you already know the id of the service.
+    ///
+    /// You should typically use [`Self::builder()`] unless you receive the
+    /// service name externally from a user (say, on the command line as an
+    /// argument).
+    pub fn new(client: Docker, id: String, attach_stdout: bool, attach_stderr: bool) -> Self {
+        Self {
+            client,
+            id,
+            attach_stdout,
+            attach_stderr,
+        }
+    }
+
+    /// Runs a service and waits for the task execution to end.
+    pub async fn run(&self, started: impl FnOnce()) -> Result<Output> {
+        let (container_id, exit_code) = loop {
+            debug!("polling tasks for service `{id}`", id = self.id);
+
+            // Get the list of tasks for the service (there should be only one)
+            let tasks = self
+                .client
+                .list_tasks(Some(ListTasksOptions {
+                    filters: HashMap::from_iter([("service", vec![self.id.as_str()])]),
+                }))
+                .await
+                .map_err(Error::Docker)?;
+
+            if tasks.is_empty() {
+                // A task hasn't been created for the service yet, query again after a delay
+                sleep(Duration::from_millis(100)).await;
+                continue;
+            }
+
+            assert_eq!(
+                tasks.len(),
+                1,
+                "Docker service task count should always be 1"
+            );
+
+            let task = tasks.into_iter().next().unwrap();
+            let status = task
+                .status
+                .expect("Docker daemon reported a task with no status");
+
+            match status.state {
+                Some(TaskState::NEW)
+                | Some(TaskState::PENDING)
+                | Some(TaskState::ALLOCATED)
+                | Some(TaskState::ASSIGNED)
+                | Some(TaskState::ACCEPTED)
+                | Some(TaskState::READY)
+                | Some(TaskState::PREPARING)
+                | None => {
+                    debug!("task has not yet started for service `{id}`", id = self.id);
```

Comment:
nit: what do you think about bumping something like this to `trace`? It just feels a bit noisy for me in `debug!`.